terraform state mv panic

### Terraform Version

```
Terraform v0.12.24
+ provider.archive v1.3.0
+ provider.aws v2.69.0
+ provider.cloudflare v2.6.0
+ provider.local v1.4.0
+ provider.template v2.1.2
```

Updated to the current major version without success:

```
Terraform v0.12.28
+ provider.archive v1.3.0
+ provider.aws v2.69.0
+ provider.cloudflare v2.6.0
+ provider.local v1.4.0
+ provider.template v2.1.2
```

### Terraform Configuration Files
[crash.log](https://github.com/hashicorp/terraform/files/4891851/crash.log)

### Steps to Reproduce

I first moved a map of zones to a non existent resource:

```
terraform state mv 'cloudflare_zone.zones' 'module.cloudflare.cloudflare_zone.zones'
```

Then I tried to correct the mistake I made and terraform panics.

```
terraform state mv 'module.cloudflare.cloudflare_zone.zones' 'module.cloudflare.cloudflare_zone.this'
```

I was able to reproduce:

1. Create a map of resources:

```
locals {
  items = toset(["A", "B", "C"])
}

resource "null_resource" "foo" {
  for_each = local.items
}
```
2. Move this configuration to a module

```
module "bug" {
  source = "./bug"
}
```

3. Move the resources to a non-existent address in the module:

```
$ terraform state mv null_resource.foo module.bug.null_resource.bar
Acquiring state lock. This may take a few moments...
Move "null_resource.foo" to "module.bug.null_resource.bar"
Successfully moved 1 object(s).
Releasing state lock. This may take a few moments...  
```

4. Move the resources to the correct address:

```
$ terraform state mv  module.bug.null_resource.bar module.bug.null_resource.foo
Acquiring state lock. This may take a few moments...
Move "module.bug.null_resource.bar" to "module.bug.null_resource.foo"
Releasing state lock. This may take a few moments...
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x197b964]
...
```

This should let you reproduce the bug.

--- a/command/state_mv.go
+++ b/command/state_mv.go
@@ -190,10 +190,7 @@ func (c *StateMvCommand) Run(args []string) int {
 				return 1
 			}
 			diags = diags.Append(c.validateResourceMove(addrFrom, addrTo))
-			if stateTo.Module(addrTo.Module) == nil {
-				// moving something to a mew module, so we need to ensure it exists
-				stateTo.EnsureModule(addrTo.Module)
-			}
+
 			if stateTo.Resource(addrTo) != nil {
 				diags = diags.Append(tfdiags.Sourceless(
 					tfdiags.Error,
@@ -223,7 +220,7 @@ func (c *StateMvCommand) Run(args []string) int {
 
 				// Update the address before adding it to the state.
 				rs.Addr = addrTo
-				stateTo.Module(addrTo.Module).Resources[addrTo.Resource.String()] = rs
+				stateTo.EnsureModule(addrTo.Module).Resources[addrTo.Resource.String()] = rs
 			}
 
 		case addrs.AbsResourceInstance:

--- a/command/state_mv_test.go
+++ b/command/state_mv_test.go
@@ -1196,6 +1196,62 @@ func TestStateMv_fromBackendToLocal(t *testing.T) {
 	testStateOutput(t, statePath, testStateMvOriginal_backend)
 }
 
+// This test covers moving the only resource in a module to a new address in
+// that module, which triggers the maybePruneModule functionality. This caused
+// a panic report: https://github.com/hashicorp/terraform/issues/25520
+func TestStateMv_onlyResourceInModule(t *testing.T) {
+	state := states.BuildState(func(s *states.SyncState) {
+		s.SetResourceInstanceCurrent(
+			addrs.Resource{
+				Mode: addrs.ManagedResourceMode,
+				Type: "test_instance",
+				Name: "foo",
+			}.Instance(addrs.IntKey(0)).Absolute(addrs.RootModuleInstance.Child("foo", addrs.NoKey)),
+			&states.ResourceInstanceObjectSrc{
+				AttrsJSON: []byte(`{"id":"bar","foo":"value","bar":"value"}`),
+				Status:    states.ObjectReady,
+			},
+			addrs.AbsProviderConfig{
+				Provider: addrs.NewLegacyProvider("test"),
+				Module:   addrs.RootModule,
+			},
+		)
+	})
+
+	statePath := testStateFile(t, state)
+	testStateOutput(t, statePath, testStateMvOnlyResourceInModule_original)
+
+	p := testProvider()
+	ui := new(cli.MockUi)
+	c := &StateMvCommand{
+		StateMeta{
+			Meta: Meta{
+				testingOverrides: metaOverridesForProvider(p),
+				Ui:               ui,
+			},
+		},
+	}
+
+	args := []string{
+		"-state", statePath,
+		"module.foo.test_instance.foo",
+		"module.foo.test_instance.bar",
+	}
+	if code := c.Run(args); code != 0 {
+		t.Fatalf("bad: %d\n\n%s", code, ui.ErrorWriter.String())
+	}
+
+	// Test it is correct
+	testStateOutput(t, statePath, testStateMvOnlyResourceInModule_output)
+
+	// Test we have backups
+	backups := testStateBackups(t, filepath.Dir(statePath))
+	if len(backups) != 1 {
+		t.Fatalf("bad: %#v", backups)
+	}
+	testStateOutput(t, backups[0], testStateMvOnlyResourceInModule_original)
+}
+
 const testStateMvOutputOriginal = `
 test_instance.baz:
   ID = foo
@@ -1512,4 +1568,24 @@ test_instance.baz:
   provider = provider["registry.terraform.io/-/test"]
   bar = value
   foo = value
+`
+
+const testStateMvOnlyResourceInModule_original = `
+<no state>
+module.foo:
+  test_instance.foo.0:
+    ID = bar
+    provider = provider["registry.terraform.io/-/test"]
+    bar = value
+    foo = value
+`
+
+const testStateMvOnlyResourceInModule_output = `
+<no state>
+module.foo:
+  test_instance.bar.0:
+    ID = bar
+    provider = provider["registry.terraform.io/-/test"]
+    bar = value
+    foo = value
 `

